src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotOverloadOperatorEqualsOnReferenceTypes.cs

```diff
@@ -0,0 +1,71 @@
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesMessage), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotOverloadOperatorEqualsOnReferenceTypesDescription), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        public static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Design,
+            RuleLevel.Disabled,
+            description: s_localizableDescription,
+            isPortedFxCopRule: true,
+            isDataflowRule: false,
+            isEnabledByDefaultInFxCopAnalyzers: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(context =>
+            {
+                var iequatableType = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIEquatable1);
+
+                context.RegisterSymbolAction(context =>
+                {
+                    var method = (IMethodSymbol)context.Symbol;
+
+                    // Method is not operator equality or not on a class
+                    if (method.MethodKind != MethodKind.UserDefinedOperator ||
+                        method.Name != "op_Equality" ||
+                        method.ContainingType.TypeKind != TypeKind.Class)
+                    {
+                        return;
+                    }
+
+                    // There's a CONSIDER rule for overriding op_Equality for anything that implements IEquatable.
+                    if (method.ContainingType.Inherits(iequatableType))
+                    {
+                        return;
+                    }
+
+                    // FxCop compat: only analyze externally visible symbols by default.
+                    if (!method.MatchesConfiguredVisibility(context.Options, Rule, context.Compilation, context.CancellationToken))
+                    {
+                        return;
+                    }
+
+                    context.ReportDiagnostic(method.CreateDiagnostic(Rule, method.ContainingType.Name));
```

Comment:
Legacy FxCop actually has couple more additional bail out conditions for this rule:

1. It does not report on `IComparable` types
2. It does not report on types that override `System.Object.Equals(object)`

We should do the same given this is a ported FxCop rule.

Reply:
Out of curiosity, are you accessing the legacy fxcop source code? Is it internal only? When trying to implement such rule it's usually a big pain for me to try and think about all cases that could (or not) lead to an issue being raised or being suppressed.

Reply:
> Is it internal only?

Unfortunately, yes. Feel free to ping me on gitter if you are considering porting any legacy FxCop rule and would like to know about the legacy implementation.